### DISCUSSION TITLE
Send server message to a specific socket

### DISCUSSION THREAD

**Discussion by jonataslaw** [uNetworking/uWebSockets.js#268] [Q&A] [answered]
Initially congratulations again for the project.
I saw that I added unsubscribeAll in the last update, and this is amazing.
Is there a possibility to make a "translator" of the functions of socket.io for uWebsockets?
I think that would help a lot. I believe that 99% of the people who end up here, are leaving Socket.io.
I saw an article about sub / pub vs join from Socket.io, but some things are still a little blurry for me.
How to send a message from the server to a specific socket?
io.to (socket).emit('message') does not appear to have a replacement.
Also with socket.io it is possible to send a message to a specific socket of the client itself with socket.to(socketReceiver).emit('message'), but the "to" method does not seem to be present.
Socket.io also has Next() which is used to confirm authentication. If the authentication result is true, it is connected, otherwise, it does not connect at all, so in case the connection goes down, it is able to communicate with the partner after reconnection, which is not the case with normal websockets (because from reconnection , those who fell receive another socket, which can send messages to the partner socket, but will not receive the messages). In addition to this, there are other approaches, in which the socket is connected, and in 1 second it must send the auth, otherwise, it is disconnected, and if the validation fails, it is also disconnected. There are no examples of auth in the repository, nor how to send messages to a specific socket, I think this would be of great help to anyone trying to migrate to the plugin.
Thanks.

**Comment by ghost** [comment]
Ws.send ?

**Comment by jonataslaw** [comment]
> Ws.send ?

Ws.send is similar Socket.emit
socket.to(socket.id).emit hasn't similar (or at least I don't know).

It is often necessary to send a message from the server to a specific client (for example, to warn that the socket on the other side has been disconnected, or that the payment method has been approved, or that the product is on its way).
Sending this to all clients can be a security issue, depending on the type of information that spreads.
Perhaps by making a loop it is possible to create a similar feature, but it is not ideal.

**Comment by ghost** [comment]
I really don't understand the issue. ws.send sends only to ws. Where ws is a WebSocket. I don't understand the problematics

**Comment by yovanoc** [comment, marked as answer]
Oh I get it, he's thinking that ws.send is like socket.emit but no, socket.emit is more similar to ws.publish and socket.to(socket.id).emit is the ws.send

**Reply by kronos-999** [reply]
Then again, ws.send does not let you specify an ID or anything similar, so how would e.g. Private Messaging be achieved?

**Reply by ghost** [reply]
ws.send is two parts; an object (the "id") and a method (send). You're complicating things, that's literally all there is. You get the ws object in open event, so it represents one individual connection.

**Comment by jonataslaw** [comment]
> I really don't understand the issue. ws.send sends only to ws. Where ws is a WebSocket. I don't understand the problematics

Sorry, I use socket.io for life, and honestly, I don't know how things work without the wrapper, there are almost no articles on the internet about uws, although socket.io 2.0 uses uws. So I suggested an excerpt on the most common / socket.io-like methods in the readme or examples, for people who want to use this lib, as well as their article on join / subs that brought me to this library which was very useful.

**Comment by jonataslaw** [comment]
> Oh I get it, he's thinking that ws.send is like socket.emit but no, socket.emit is more similar to ws.publish and socket.to(socket.id).emit is the ws.send

Thank you, that was what I needed to know!
